### ISSUE TITLE
docker module always reload container if net parameter ommited

### ISSUE THREAD

**Issue by awasilyev** [ansible/ansible-modules-core#1885]
##### Ansible Configuration:

[defaults]
hash_behaviour=merge
host_key_checking=false
gathering=no
pipelining=yes
remote_user=centos
become=yes
sudo=yes
hostfile=hosts
forks=30
##### Environment:

CentOS 7
##### Summary:

docker module always reload container if net parameter ommited
##### Steps To Reproduce:

[centos@dpt-bb0 ~]$ ansible localhost -c local -s -m docker -a "image=mongo state=reloaded"
localhost | success >> {
    "ansible_facts": {
        "docker_containers": [
            {
                "Id": "8a125c823d5de6d7fe8d4338f5b68c2324c30ad7a1bbe85023b30f0af242bb3f",
                "Warnings": null
            }
        ]
    },
    "changed": true,
    "containers": [
        {
            "Id": "8a125c823d5de6d7fe8d4338f5b68c2324c30ad7a1bbe85023b30f0af242bb3f",
            "Warnings": null
        }
    ],
    "msg": "started 1 container, created 1 container, pulled 1 container.",
    "reload_reasons": null,
    "summary": {
        "created": 1,
        "killed": 0,
        "pulled": 1,
        "removed": 0,
        "restarted": 0,
        "started": 1,
        "stopped": 0
    }
}

[centos@dpt-bb0 ~]$ ansible localhost -c local -s -m docker -a "image=mongo state=reloaded"
localhost | success >> {
    "ansible_facts": {
        "docker_containers": [
            {
                "Id": "5130248d943ae9485e6287caa601d10a7bb93477df543f5e8b9db145db25da39",
                "Warnings": null
            }
        ]
    },
    "changed": true,
    "containers": [
        {
            "Id": "5130248d943ae9485e6287caa601d10a7bb93477df543f5e8b9db145db25da39",
            "Warnings": null
        }
    ],
    "msg": "removed 1 container, stopped 1 container, started 1 container, created 1 container.",
    "reload_reasons": "net (bridge => )",
    "summary": {
        "created": 1,
        "killed": 0,
        "pulled": 0,
        "removed": 1,
        "restarted": 0,
        "started": 1,
        "stopped": 1
    }
}
##### Expected Results:

Second run should not restart container 
##### Actual Results:

Container always restarted

**Comment by bcoca** [comment]
Can You Help Us Out?
====================

Thanks for filing a ticket! I am the friendly GitHub Ansibot.

It looks like you might not have filled out the issue description based on our standard issue template.  You might not have known about that, and that's ok too, we'll tell you how to do it.

We have a standard template because Ansible is a really busy project and it helps to have some standard information in each ticket, and GitHub doesn't yet provide a standard facility to do this like some other bug trackers.  We hope you understand as this is really valuable to us!.

Solving this is simple: please copy the contents of this [template](https://raw.githubusercontent.com/ansible/ansible/devel/ISSUE_TEMPLATE.md) and **paste it into the description** of your ticket.  That's it!

If You Had A Question To Ask Instead
====================================

If you happened to have a "how do I do this in Ansible" type of question, that's probably more of a user-list question than a bug report, and you should probably ask this question on the project mailing list instead.

However, if you think you have a bug, the report is the way to go!  We definitely want all the bugs filed :)  Just trying to help!

About Priority Tags
===================

Since you're here, we'll also share some useful information at this time.

In general tickets will be assigned a priority between P1 (highest) and P5, and then worked in priority order. We may also have some follow up questions along the way, so keeping up with follow up comments via GitHub notifications is a good idea.

Due to large interest in Ansible, humans may not comment on your ticket immediately.

Mailing Lists
=============

If you have concerns or questions, you're welcome to stop by the ansible-project or ansible-development mailing lists, as appropriate.  Here are the links:

  * https://groups.google.com/forum/#!forum/ansible-project - for discussion of bugs and how-to type questions
  * https://groups.google.com/forum/#!forum/ansible-devel - for discussion on how to implement a code change, or feature brainstorming among developers

Thanks again for the interest in Ansible!

**Comment by awasilyev** [comment]
##### Issue Type: Bug Report

##### Ansible Version: 1.9.1

##### Ansible Configuration:
[defaults]
hash_behaviour=merge
host_key_checking=false
gathering=no
pipelining=yes
remote_user=centos
become=yes
sudo=yes
hostfile=hosts
forks=30

##### Environment:
CentOS 7

##### Summary:
docker module always reload container if net parameter ommited

##### Steps To Reproduce:

[centos@dpt-bb0 ~]$ ansible localhost -c local -s -m docker -a "image=mongo state=reloaded"
localhost | success >> {
    "ansible_facts": {
        "docker_containers": [
            {
                "Id": "8a125c823d5de6d7fe8d4338f5b68c2324c30ad7a1bbe85023b30f0af242bb3f",
                "Warnings": null
            }
        ]
    },
    "changed": true,
    "containers": [
        {
            "Id": "8a125c823d5de6d7fe8d4338f5b68c2324c30ad7a1bbe85023b30f0af242bb3f",
            "Warnings": null
        }
    ],
    "msg": "started 1 container, created 1 container, pulled 1 container.",
    "reload_reasons": null,
    "summary": {
        "created": 1,
        "killed": 0,
        "pulled": 1,
        "removed": 0,
        "restarted": 0,
        "started": 1,
        "stopped": 0
    }
}

[centos@dpt-bb0 ~]$ ansible localhost -c local -s -m docker -a "image=mongo state=reloaded"
localhost | success >> {
    "ansible_facts": {
        "docker_containers": [
            {
                "Id": "5130248d943ae9485e6287caa601d10a7bb93477df543f5e8b9db145db25da39",
                "Warnings": null
            }
        ]
    },
    "changed": true,
    "containers": [
        {
            "Id": "5130248d943ae9485e6287caa601d10a7bb93477df543f5e8b9db145db25da39",
            "Warnings": null
        }
    ],
    "msg": "removed 1 container, stopped 1 container, started 1 container, created 1 container.",
    "reload_reasons": "net (bridge => )",
    "summary": {
        "created": 1,
        "killed": 0,
        "pulled": 0,
        "removed": 1,
        "restarted": 0,
        "started": 1,
        "stopped": 1
    }
}

##### Expected Results:

Second run should not restart container 

##### Actual Results:

Container always restarted

**Comment by leedm777** [comment]
I just hit this with `state: reloaded`. Adding `net: default` to the task works around the problem.

**Comment by leedm777** [comment]
Confirmed still an issue in v1.9.3-0.1.rc1

**Comment by mrwacky42** [comment]
**sigh**
_insert beleaguered cry for some sort of idempotent operation here_

**Comment by ryanwalls** [comment]
Yep, I'm seeing this problem as well.  Version 1.9.4.

**Comment by mrwacky42** [comment]
#2215 was merged into the stable-1.9 branch, so a fix would presumably be in 1.9.5

**Comment by joshmoore** [comment]
Same problem in `v2.0.0-0.6.rc1`; https://github.com/ansible/ansible-modules-core/issues/1885#issuecomment-130464795 works.

**Comment by jmnas** [comment]
I see this with `v2.0.0-0.6.rc1` as well. The problem is that #2215 was merged into `stable-1.9` but is not in `stable-2.0`. That branch has 29e4a127 but that fix doesn't work for me. If I make the changes in #2215 it does work correctly for me.

I'm running Ubuntu 15.10, Docker version 1.9.1, and docker-py 1.6.0.

**Comment by bcoca** [comment]
@jmnas let me know if the PR above fixes it for you

**Comment by jmnas** [comment]
@bcoca, #2612 does fix it in the environment above. However, I just tested in another one I have (RHEL 6, Docker 1.7, docker-py 1.4) and it actually breaks there, it seems `bridge` is what is needed there. Might be a docker 1.7 vs. docker 1.9 thing?

**Comment by bcoca** [comment]
hmm, seems like we need version detection, anyone know the exact version in which this changes?

i'll look it up in the morning.

**Comment by jmnas** [comment]
I think docker-py made the change with this: docker/docker-py@8f1a82f so 1.20? I don't actually see anything about that in the [1.20 docker remote API docs](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.20/) though.

**Comment by ryanwalls** [comment]
Still seeing this issue in ansible 2.0.0.2 and Docker 1.10.  The reload reason is: 

```
"reload_reasons": "net (default => bridge)"
```

**Comment by dkerwin** [comment]
Works in ansible 2.0.1.0 with setting net: default

**Comment by elouanKeryell-Even** [comment]
I've also run into this issue with ansible 2.0 and docker 1.10.

**Comment by BirgerK** [comment]
Problem still existing with Ansible 2.1.

Named workaround still does work:

```
net: default
```

**Comment by ansibot** [comment]
@cove, @softzilla, @zfil, @ThomasSteinbach, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by jctanner** [comment]
!needs_info

**Comment by ansibot** [comment]
@cove, @softzilla, @zfil, @ThomasSteinbach, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by jctanner** [comment]
The "docker" module is deprecated and ansible 1.9 is no longer supported.
